src/StripeTests/Functional/TelemetryTest.cs

```diff
@@ -0,0 +1,216 @@
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Once(),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        TelemetryHeaderMatcher(
+                            m.Headers,
+                            (s) => s == "req_1",
+                            (d) => d >= 10)),
+                    ItExpr.IsAny<CancellationToken>());
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Once(),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        TelemetryHeaderMatcher(
+                            m.Headers,
+                            (s) => s == "req_2",
+                            (d) => d >= 20)),
+                    ItExpr.IsAny<CancellationToken>());
+        }
+
+        [Fact]
+        public async Task TelemetryWorksWithConcurrentRequests()
+        {
+            this.ResetStripeClient();
+
+            var invocationCount = 0;
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Setup<Task<HttpResponseMessage>>(
+                    "SendAsync",
+                    ItExpr.IsAny<HttpRequestMessage>(),
+                    ItExpr.IsAny<CancellationToken>())
+                .Returns(async () =>
+                    {
+                        invocationCount += 1;
+
+                        await Task.Delay(10);
+
+                        var response = new HttpResponseMessage(HttpStatusCode.OK);
+                        response.Headers.Add("Request-Id", $"req_{invocationCount}");
+                        response.Content = new StringContent("{}");
+                        return response;
+                    });
+
+            StripeConfiguration.EnableTelemetry = true;
+            var service = new BalanceService();
+
+            // the first 2 requests will not contain telemetry
+            await Task.WhenAll(service.GetAsync(), service.GetAsync());
+
+            // the following 2 requests will contain telemetry
+            await Task.WhenAll(service.GetAsync(), service.GetAsync());
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Exactly(2),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        !m.Headers.Contains("X-Stripe-Client-Telemetry")),
+                    ItExpr.IsAny<CancellationToken>());
+
+            this.MockHttpClientFixture.MockHandler.Protected()
+                .Verify(
+                    "SendAsync",
+                    Times.Exactly(2),
+                    ItExpr.Is<HttpRequestMessage>(m =>
+                        TelemetryHeaderMatcher(
+                            m.Headers,
+                            (s) => s == "req_1" || s == "req_2",
```

Comment:
It would be nice if could add all of the `request_id`s that the MockHandler sees to a set, and then asserting that the size of the set is 2 here. That way, instead of asserting that the sent metrics were "req_1" or "req_2", we assert that _both_ "req_1" and "req_2" were sent in some order.

Reply:
Ack. I can't think of an easy way to do that, but given that there are only 2 requests, I'll just duplicate the `Verify()` to check that there was one invocation with "req_1" and another with "req_2". Not the most scalable solution, but I think it's sufficient for the purpose of this test.